### ISSUE TITLE
The SHOPT_FIXEDARRAY feature code should be removed

### ISSUE THREAD

**Issue by krader1961** [att/ast#234]
The *src/cmd/ksh93/RELEASE* file says that fixed size arrays (compile time symbol `SHOPT_FIXEDARRAY`) was disabled on 2012-09-12. But the code at the time it was open sourced on 2016-01-11 shows that the feature was enabled. There is no comment anywhere explaining why it was re-enabled. I suspect someone did so solely to continue working on the feature.

Note that at this time the implementation appears to be incomplete. As evidenced by the `#if xSHOPT_FIXEDARRAY` tests (notice the leading `x`) in *src/cmd/ksh93/sh/nvtree.c* and the `#undef SHOPT_FIXEDARRAY` at the top of *src/cmd/ksh93/sh/arith.c*.

We should not be building ksh93 with this feature enabled. In fact, I think it should be removed since it is incomplete after more than six years of work, there are no unit tests for it, and it appears to be undocumented. It also isn't present in the last stable release (93u) used by every distro so there shouldn't be anyone depending on it.

**Comment by krader1961** [comment]
Ugh! Notice that the *src/cmd/ksh93/README* file lists several compile time symbols that are either no longer relevant or whose current state is different than documented in that file. An example of the latter is `SHOPT_KIA` which that file says is disabled by default when in fact it is enabled. There are others, such as `SHOPT_OO`, which are no longer present in the code.

I hate to throw away work that may have value but I am sorely tempted to remove not only the code that depends on `SHOPT_FIXEDARRAY` but also several other symbols such as `SHOPT_KIA`. If they haven't become standard features by now that are documented and have unit tests the code should probably be dropped.

**Comment by krader1961** [comment]
I've looked more closely at a few of these symbols. `SHOPT_KIA` for example seems to date back to 1997 and the code it guards should just be included unconditionally. It may be that `SHOPT_FIXEDARRAY` is the only feature that guarded by a build time symbol that isn't ready for prime time.

**Comment by krader1961** [comment]
I've removed this code with commit 54cd98f2d.

**Comment by krader1961** [comment]
In light of the discussion in issue #240 I'm going to revert the removal of this "feature". The scare quotes are intentional since the feature has no unit tests or documentation. Furthermore, as I wrote in commit 54cd98f2d:

```
    This feature has been in development for more than seven years. It is
    not present in any released version of `ksh`. It has no unit tests.
    Parts of the code for this function isn't even compiled. Remove it.
```

But several individuals have expressed concern about removing this "feature". So I'm just going to reinstate it while removing the `SHOPT_FIXEDARRAY` so that it is enabled unconditionally. It will be interesting to see how many problems get attributed to this "feature". Yay for democracy!

P.S., TBD is what to do about the block of code predicated on `#if xSHOPT_FIXEDARRAY` which is never compiled.

**Comment by krader1961** [comment]
@kdudka, Do you agree the "SHOPT_FIXEDARRAY" feature should be reinstated?

**Comment by dannyweldon** [comment]
I don't think it should be made unconditional yet if it is still experimental and not even working properly.  I have not tested it out myself, but would certainly be willing to build `ksh` with it enabled to help out with testing.  If we can write some tests to prove that it works, then maybe it is ready to be released into the wild.

**Comment by krader1961** [comment]
@dannyweldon, That's a reasonable point. The problem is that the manner in which this feature is integrated with the rest of the code all but guarantees mistakes will be made. That's because the conditional inclusion results in hard to understand constructs like

```c
if (something) {
    // whatever
} else
#if SHOPT_FIXEDARRAY
{
    // long block of code
}
#else
    // single statement
#endif
```

This is true for many of the other conditionally compiled features. That type of code either needs to be included unconditionally, removed, or refactored into clean abstractions with helper functions, dispatch tables, or some similar mechanism.

I reopened this because there has been much wailing and gnashing of teeth over the removal of any feature including this one. With multiple individuals effectively saying "who are you to make such a decision." To which my reply is I'm one of very few people actually contributing changes in the hope of making a post ksh93u+ version relevant (the beta branch is ksh93v-). And I have zero interest in trying to finish implementing this feature and documenting it.

**Comment by kdudka** [comment]
On Friday, January 19, 2018 6:36:18 AM CET Kurtis Rader wrote:
> @kdudka, Do you agree the "SHOPT_FIXEDARRAY" feature should be reinstated?

I will delegate the decision to @siteshwar.  For Red Hat, it is mainly 
important to have as fewer user-visible changes as possible, compared to
the latest RHEL-7 ksh package, which is based on ast-ksh.2012-08-01.tgz.

If the feature was introduced after that release, our users/customers will not 
notice the removal.  Similarly, if we compiled it with SHOPT_FIXEDARRAY turned 
off, its removal will not result in a user-visible change.

On the other hand, if someone compiled ksh with that feature disabled and it 
became unconditionally enabled now, it might result in some user-visible side 
effects, even when the feature itself stays unused in the target scripts.

**Comment by siteshwar** [comment]
`SHOPT_FIXEDARRAY` option was rather new. Here is related changelog from `RELEASE` file when it was introduced :

```
10-09-28 +A new compile option SHOPT_FIXEDARRAY has been added and is being
      evaluation.  It allows fixed sized indexed arrays be to defined
      using "typeset array[dim1][dim2]...[dimn]".  Fixed sized arrays
      are used the same way indexed arrays are.  Currently, only fixed
      arrays of fixed objects (float, int, and justifies objects) are
      supported.
```

It was enabled in last stable release, but was disabled by default later :

```
12-09-12  Changed the default build to disable SHOPT_FIXEDARRAY.
```

Since this feature was new, it is unlikely that any of RHEL users would be using it. So legacy scripts should not depend on it. My only concern is that this option was enabled by default, so there is a possibility that disabling/removing this code may have side effects on other features. 

@krader1961 If you have removed this code, I would suggest you not to reinstate it now. But while debugging any regressions, we should keep in mind that this code was removed.

**Comment by krader1961** [comment]
AFAICT removing this code had no effect on any unit test.

**Comment by krader1961** [comment]
FWIW, A lot of test failures when building with ASAN enabled involve code adjacent to the blocks protected by `#if SHOPT_FIXEDARRAY`. I removed that code before I added support for ASAN builds. I'm going to reinstate that code in a test branch to see if doing so reduces the number of failures when ASAN is enabled. I'm hoping the answer is zero but will not be surprised if it is non-zero.

**Comment by krader1961** [comment]
It's not obvious that reinstating the code protected by `#if SHOPT_FIXEDARRAY` is going to be useful. While doing so I've already found a bug in that code that was introduced after the ksh93u release. In the ksh93v code the `scope()` function in *src/cmd/ksh93/sh/arith.c* looks like this:

```
#if SHOPT_FIXEDARRAY
                if(n)
                {
                        Namarr_t *ap = nv_arrayptr(np);
                        ap->nelem = dim;
                        nv_putsub(np,(char*),n,0);
                }
                else
#endif /* SHOPT_FIXEDARRAY */
```

That fails to compile due to the syntax error in the `nv_putsub()` call. It looked like this in the ksh93u code (note the placement of the `0`):

```
     nv_putsub(np,(char*)0,n);
```

Yes, it's a trivial typo. If that's the only such error maybe my effort to see if reinstating that code will be useful in answering the question whether that code affects the ASAN use-after-free errors.

**Comment by krader1961** [comment]
Sigh. Reinstating the `SHOPT_FIXEDARRAY` code isn't feasible as there have been too many changes in the 14 months since I ripped it out.

I made a half-hearted attempt to enable ASAN in the commit prior to removing the `SHOPT_FIXEDARRAY` code but was not successful. That's because every single unit test fails with a "global-buffer-overflow" error. Which is one of the bugs I fixed long ago and I'm not going to backport that fix. Especially since I've fixed so many other memory management bugs in the interim that it's not clear this effort would answer the question of whether removing the `SHOPT_FIXEDARRAY` code is why any of the remaining "use-after-free" bugs exist.
